unit-tests/native/src/test/scala/java/lang/ref/WeakReferenceTest.scala

```diff
@@ -0,0 +1,96 @@
+package java.lang.ref
+
+import java.lang.ref._
+
+import org.junit.Test
+import org.junit.Assert._
+import org.junit.Assume._
+
+import scala.scalanative.runtime.GCInfo
+import scala.scalanative.runtime.GCInfo._
+
+// "AfterGC" tests are very sensitive to optimizations,
+// both by Scala Native and LLVM.
+class WeakReferenceTest {
+
+  case class A()
+
+  def gcAssumption(): Unit = {
+    assumeTrue(
+      "WeakReferences work only on Commix and Immix GC",
+      GCInfo.getType() == Commix() || GCInfo.getType() == Immix()
+    )
+  }
+
+  @noinline def allocWeakRef(
+      referenceQueue: ReferenceQueue[A]
+  ): WeakReference[A] = {
+    var a = A()
+    val weakRef = new WeakReference(a, referenceQueue)
+    assertEquals("get() should return object reference", weakRef.get(), A())
+    a = null
+    weakRef
+  }
+
+  @Test def addsToReferenceQueueAfterGC(): Unit = {
+    gcAssumption()
+    val refQueue = new ReferenceQueue[A]()
+    val weakRef1 = allocWeakRef(refQueue)
+    val weakRef2 = allocWeakRef(refQueue)
+    val weakRefList = List(weakRef1, weakRef2)
+
+    // limit - arbitrary number that can be higher or lower,
+    // depending on how much garbage is necessary to perform GC
+    val limit = 1000000
+    for (i <- 0 to limit) {
+
+      // Allocation spamming. GC.collect() proved to be unreliable
+      // for commix (as in issue #2367), so we force garbage collection
+      // to happen organically.
+      val data = A()
+      if (data.toString() == "a") {
+        println(data)
```

Comment:
Here I was trying to outsmart the optimizer, this println is actually unreachable, but it gives purpose to `data` and more importantly `data.toString()` - our otherwise purposeless memory leaker. I'll try to simplify this (or at least add comments)

Reply:
Probably annotating a method with `nooptimize` should make sure that the optimizer would not try to dead eliminate the code.